Upgrade RouteProgress Step Data

danesfeder opened the pull request:
Closes #325 

In preparation for a tunnel and dead reckoning work, we need to expose more data in the `RouteProgress` object.   This PR does this and updates the tests accordingly

devotaaabel commented on libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/BaseTest.java:
```diff
@@ -27,4 +33,12 @@ protected String loadJsonFixture(String filename) throws IOException {
     return scanner.hasNext() ? scanner.next() : "";
   }
+
+  protected LegStep getFirstStep(DirectionsRoute route) {
```
Is there somewhere else these could go? It seems very specific to be included in all tests

devotaaabel commented on libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteStepProgress.java:
```diff
@@ -25,29 +26,118 @@
+      double distanceRemaining = distanceRemaining();
+      double distanceTraveled = calculateDistanceTraveled(step, distanceRemaining);
+      distanceTraveled(distanceTraveled);
+      float fractionTraveled = calculateFractionTraveled(step, distanceTraveled);
```
Why are we calculating all of this before it's needed?

danesfeder commented:
@Guardiola31337 @devotaaabel tests finished!  This is good to go 🚀 for 👀

devotaaabel commented on libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/BaseTest.java:
```diff
@@ -27,4 +42,55 @@ protected String loadJsonFixture(String filename) throws IOException {
     return scanner.hasNext() ? scanner.next() : "";
   }
+
+  protected RouteProgress buildRouteProgress(DirectionsRoute route,
```
I think instead of putting this in `BaseTest` we should have classes that fake data for us, like `RouteProgressFaker`, just so we don't clutter up `BaseTest` with implementation specific code

Guardiola31337 replied:
@danesfeder 
> I extracted the building logic from `BaseTest` into classes that separate the building concerns.

How did you address @devotaaabel's comment? `buildRouteProgress` is still in `BaseTest` 🤔 Could you clarify?

Guardiola31337 commented on libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java:
```diff
@@ -28,10 +34,21 @@
+  private double stepDistanceRemaining;
   private boolean shouldIncreaseIndex;
 
   NavigationRouteProcessor() {
```
Although theses changes have refactored `NavigationRouteProcessor` there are still some methods kinda long. What about extracting blocks of code into well-named (based on the comments if there are any) `private` methods and simplifying conditional expressions?

Guardiola31337 commented on libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/offroute/OffRouteDetector.java:
```diff
@@ -24,7 +23,6 @@
   private OffRouteCallback callback;
   private RingBuffer<Integer> distancesAwayFromManeuver = new RingBuffer<>(3);
-  private List<Point> stepPoints = new ArrayList<>();
 
```
Same here ☝️

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/BaseTest.java:
```diff
@@ -34,6 +42,7 @@
 
   public static final String ACCESS_TOKEN = "pk.XXX";
   private static final String DIRECTIONS_PRECISION_6 = "directions_v5_precision_6.json";
+  private static final int FIRST_POINT = 0;
```
👀
> Private field 'FIRST_POINT' is never used

The same thing happens with `compareJson` and `getFirstLeg` 👇

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/BaseTest.java:
```diff
@@ -49,16 +58,55 @@ protected String loadJsonFixture(String filename) throws IOException {
-      .legIndex(0)
-      .build();
+    return buildRouteProgress(aRoute, 100, 100,
+      100, 0, 0);
```
NIT unnecessary new line

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/milestone/StepMilestoneTest.java:
```diff
@@ -28,21 +24,17 @@
 
   @Before
-  public void setup() throws IOException {
+  public void setup() throws Exception {
```
In general, we should avoid inline setup 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=65
Could you extract this block of code into private "factory" methods called and use them only in the tests that actually need it/across the rest of the tests?
IMO, it reflects better what's going on in the test and you don't break the [triple A](https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40) "rule".

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/milestone/StepMilestoneTest.java:
```diff
@@ -15,9 +14,6 @@
-
-import java.io.IOException;
 
 @RunWith(RobolectricTestRunner.class)
```
Why do we need `RobolectricTestRunner`?

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/milestone/TriggerPropertyTest.java:
```diff
@@ -17,8 +17,6 @@
 
-import java.io.IOException;
-
 @RunWith(RobolectricTestRunner.class)
```
Why are we using the `RobolectricTestRunner`?

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/milestone/TriggerPropertyTest.java:
```diff
@@ -29,22 +27,18 @@
 
   @Before
-  public void setup() throws IOException {
+  public void setup() throws Exception {
```
Same here ☝️
Great opportunity to give some ❤️ to the tests as well!

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/milestone/TriggerTest.java:
```diff
@@ -35,7 +34,7 @@
 
   @Before
-  public void setup() throws IOException {
+  public void setup() throws Exception {
```
Same `RobolectricTestRunner ` and `setup` comments pointed out above.

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/FasterRouteDetectorTest.java:
```diff
@@ -116,16 +116,8 @@ public void onNullRouteProgressPassed_shouldCheckFasterRouteIsFalse() throws Exc
-
-    return defaultRouteProgress;
+    return buildRouteProgress(aRoute, 100,
+      700, 1000, 0, 0);
```
NIT unnecessary new line

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/FasterRouteDetectorTest.java:
```diff
@@ -116,16 +116,8 @@ public void onNullRouteProgressPassed_shouldCheckFasterRouteIsFalse() throws Exc
 
```
Normally is good to structure the tests following the triple A 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40 http://wiki.c2.com/?ArrangeActAssert
If the `A`rrange part gets too long we should create factory or builder methods.

This applies to other modified tests included in this PR.

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationEventDispatcherTest.java:
```diff
@@ -77,14 +75,9 @@ public void setup() throws IOException {
+    routeProgress = buildRouteProgress(route, 100, 100,
+      100, 0, 0);
   }
 
```
Generally is preferable having only 1 assertion per test 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=57 The same "rule" applies to _Behavior tests_ (`verify`).
This applies to other tests included in `NavigationEventDispatcherTest`.

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationEventDispatcherTest.java:
```diff
@@ -324,28 +317,16 @@ public void setNavigationMetricListener_didNotGetTriggeredUntilArrival() throws
     navigationEventDispatcher.addMetricArrivalListener(arrivalListener);
 
     // Progress that hasn't arrived
```
Why having comments in a test? Also 👇

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationHelperTest.java:
```diff
@@ -36,77 +45,57 @@
 
 @RunWith(RobolectricTestRunner.class)
 @Config(constants = BuildConfig.class, sdk = 25)
 public class NavigationHelperTest extends BaseTest {
```
Overall same comments pointed out above ☝️

danesfeder replied:
`RobolectricTestRunner` is needed to test the milestones being triggered - removing this results in a crash

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationHelperTest.java:
```diff
@@ -133,35 +123,344 @@ public void offRouteDetectionDisabled_isOffRouteReturnsFalse() throws Exception
   }
+
+  @Test
+  public void createIntersectionList_returnsCompleteIntersectionList() throws Exception {
```
There are some blocks of duplicated code 👇 
What about extracting them into `private` "factory" methods and used them across the tests?

danesfeder replied:
If you're talking about:
```
    RouteProgress routeProgress = buildMultiLegRouteProgress();
    LegStep currentStep = routeProgress.currentLegProgress().currentStep();
    LegStep upcomingStep = routeProgress.currentLegProgress().upComingStep();
```
This can't be extracted, as we need both current and upcoming steps for the `Arrange`

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/offroute/OffRouteDetectorTest.java:
```diff
@@ -43,7 +43,6 @@ public void setup() throws Exception {
     offRouteDetector.setOffRouteCallback(mockCallback);
-    setupStepPoints(offRouteDetector);
   }
 
```
Overall same comments pointed out above ☝️

danesfeder replied:
Not sure what the comment above here means

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteLegProgressTest.java:
```diff
@@ -17,77 +13,41 @@
-    String body = loadJsonFixture(DIRECTIONS_PRECISION_6);
-    DirectionsResponse response = gson.fromJson(body, DirectionsResponse.class);
-    route = response.routes().get(0);
+  public void setup() throws Exception {
```
Same with `RouteLegProgressTest`.

Guardiola31337 replied:
The same thing happens with `RouteProgressTest`, `RouteStepProgressTest`, `SnapToRouteTest` and `ToleranceUtilsTest`.

danesfeder commented on libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationHelper.java:
```diff
@@ -52,6 +70,17 @@ static Point userSnappedToRoutePosition(Location location, List<Point> coordinat
     return ((Point) feature.geometry());
   }
 
+  static Location buildSnappedLocation(MapboxNavigation mapboxNavigation, boolean snapToRouteEnabled,
```
Given the size of this PR already and that this is core logic we will be touching, let's save this refactor for another PR that touches this code.

Guardiola31337 replied:
Please write a reminder down so we don't forget to 👀 this comment when touching `NavigationHelper` 😬

danesfeder commented on libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationHelper.java:
```diff
@@ -240,9 +495,48 @@ static Point nextManeuverPosition(int stepIndex, List<LegStep> steps, List<Point
     return !coords.isEmpty() ? coords.get(coords.size() - 1) : coords.get(coords.size());
```
Made it package private to test it.  This will be addressed by refactoring `NavigationHelper` as a whole

danesfeder commented:
@Guardiola31337 Outside of your comments surrounding `NavigationHelper`, I believe I addressed most of your concerns / had some follow up questions.  In the spirit of our recent talks, I _don't_ think we should open a ticket, but keep it in mind moving forward.  This core logic will surely be touched soon and grant us an opportunity to clean up / remove the `NavigationHelper` class. 

@devotaaabel I extracted the building logic from `BaseTest` into classes that separate the building concerns. 

Thanks for the great review!

Guardiola31337 commented on libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/camera/DynamicCameraTest.java:
```diff
@@ -10,50 +10,39 @@
 
-@RunWith(RobolectricTestRunner.class)
-@Config(constants = BuildConfig.class, manifest = Config.DEFAULT_MANIFEST_NAME)
 public class DynamicCameraTest extends BaseTest {
```
In general is hard to know where the assert values come from. I know that we're taking the values from `"directions_v5_precision_6.json"` fixture but IMO it's not clear when you 👀 the tests at first glance. There is some kind of magic behind the scenes that forces you to 👀 the implementation to fully understand what's going on. We should remove that necessity.

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/milestone/TriggerPropertyTest.java:
```diff
@@ -11,44 +11,21 @@
 @Config(constants = BuildConfig.class)
 public class TriggerPropertyTest extends BaseTest {
 
   // Fixtures
```
NIT unnecessary comment

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationEventDispatcherTest.java:
```diff
@@ -323,29 +316,16 @@ public void setNavigationMetricListener_didNotGetTriggeredUntilArrival() throws
     verify(eventListeners, times(1)).onRouteProgressUpdate(routeProgressDidNotArrive);
     verify(arrivalListener, times(0)).onArrival(location, routeProgressDidNotArrive);
 
     // Progress that has arrived
```
NIT unnecessary comment

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteLegProgressTest.java:
```diff
@@ -117,174 +77,129 @@ public void previousStep_returnsPreviousStep() {
+    RouteLeg firstLeg = route.legs().get(0);
     double stepSegments = 5000; // meters
 
     // Chop the line in small pieces
```
What about extracting this block of code into a private method and give it a name based on the comment? This way the test will be easier to read and understand and comment will become superfluous so it won't be necessary. Same thing with `double stepSegments = 5000; // meters` ☝️

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgressTest.java:
```diff
@@ -5,142 +5,134 @@
-@Config(constants = BuildConfig.class, sdk = 21, shadows = {ShadowLocationManager.class})
 public class RouteProgressTest extends BaseTest {
 
   // Fixtures
```
NIT unnecessary comment

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteLegProgressTest.java:
```diff
@@ -117,174 +77,129 @@ public void previousStep_returnsPreviousStep() {
+    routeProgress = routeProgress.toBuilder().stepIndex(5).build();
+
     assertTrue(routeProgress.currentLegProgress().followOnStep().geometry()
       .startsWith("un`ffAz_dogFz`Aq^hF{B|GaB|HcAxKKlIp@lOzC|Dh@hKtAzFh@`FDvHy@bG{AjKaEfF"));
```
What's the value of testing geometries this way? Do geometries start always with that value? Could you clarify? If a test doesn't add any value :skull_and_crossbones:

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteLegProgressTest.java:
```diff
@@ -17,77 +12,46 @@
-      .legIndex(0)
-      .build();
     assertTrue(routeProgress.currentLegProgress().upComingStep().geometry()
       .startsWith("so{gfA~}xpgFzOyNnRoOdVqXzLmQbDiGhKqQ|Vie@`X{g@dkAw{B~NcXhPoWlRmXfSeW|U"));
```
What's the value of testing geometries this way? Do geometries start always with that value? Could you clarify? If a test doesn't add any value :skull_and_crossbones:

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgressTest.java:
```diff
@@ -149,157 +141,141 @@ public void legIndex_returnsCurrentLegIndex() {
     assertEquals(0, routeProgress.fractionTraveled(), BaseTest.LARGE_DELTA);
   }
 
   // TODO check fut
```
What's `fut`?

danesfeder replied:
Dunno lol

Guardiola31337 replied:
Let's remove it then 💀

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteStepProgressTest.java:
```diff
@@ -29,83 +28,60 @@
 
   // Fixtures
```
NIT unnecessary comment
What about adding the suffix `_FIXTURE` to the fixture constants?

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgressTest.java:
```diff
@@ -149,157 +141,141 @@ public void legIndex_returnsCurrentLegIndex() {
    */
```
What about splitting `RouteProgressTest` into two test classes, one including "normal" leg route tests whereas the other `Multi-leg route` tests? If that's important in any sense, if not we should remove the comment.

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteStepProgressTest.java:
```diff
@@ -29,83 +28,60 @@
+    RouteLeg firstLeg = route.legs().get(0);
+
     LineString lineString
       = LineString.fromPolyline(firstLeg.steps().get(5).geometry(), Constants.PRECISION_6);
```
NIT unnecessary new line

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteStepProgressTest.java:
```diff
@@ -29,83 +28,60 @@
+    LegStep firstStep = route.legs().get(0).steps().get(0);
+
     LineString lineString
       = LineString.fromPolyline(firstStep.geometry(), Constants.PRECISION_6);
```
NIT unnecessary new line

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteStepProgressTest.java:
```diff
@@ -123,53 +99,48 @@ public void distanceRemaining_equalsCorrectValueAtIntervals() {
+  public void distanceRemaining_equalsZeroAtEndOfStep() throws Exception {
+    DirectionsRoute route = buildTestDirectionsRoute();
+    RouteProgress routeProgress = buildTestRouteProgress(route, 0,
+      0, 0, 3, 0);
```
NIT unnecessary new line

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteStepProgressTest.java:
```diff
@@ -246,60 +220,61 @@ public void fractionTraveled_equalsCorrectValueAtIntervals() {
+    RouteLeg firstLeg = route.legs().get(0);
+    LegStep firstStep = route.legs().get(0).steps().get(0);
     LineString lineString
       = LineString.fromPolyline(firstStep.geometry(), Constants.PRECISION_6);
```
NIT unnecessary new line

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteStepProgressTest.java:
```diff
@@ -29,83 +28,60 @@
+
     LineString lineString
       = LineString.fromPolyline(firstStep.geometry(), Constants.PRECISION_6);
     double stepDistance = TurfMeasurement.length(lineString, TurfConstants.UNIT_METERS);
```
What about extracting `// Chop the line in small pieces` block of code into a private method and give it a name based on the comment? This way the test will be easier to read and understand and comment will become superfluous so it won't be necessary. Same thing with `double stepSegments = 5000; // meters` 👇

Also 👀 a bunch of `// Chop the line in small pieces` blocks of code in other tests, if it's duplicated code we should consolidate it and use it across the tests.

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/snap/SnapToRouteTest.java:
```diff
@@ -2,77 +2,36 @@
 import static junit.framework.Assert.assertNotNull;
 import static junit.framework.Assert.assertTrue;
 
 @RunWith(RobolectricTestRunner.class)
```
Why are we using the `RobolectricTestRunner`?

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/snap/SnapToRouteTest.java:
```diff
@@ -2,77 +2,36 @@
   @Test
   public void sanity() throws Exception {
     Snap snap = new SnapToRoute();
     assertNotNull(snap);
```
👀 https://github.com/mapbox/mapbox-navigation-android/pull/812#discussion_r182118450

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/utils/ToleranceUtilsTest.java:
```diff
@@ -24,75 +19,44 @@
+    double distanceToIntersection = route.distance() - 39;
+    LineString lineString = LineString.fromPolyline(route.geometry(), Constants.PRECISION_6);
     Point closePoint
       = TurfMeasurement.along(lineString, distanceToIntersection, TurfConstants.UNIT_METERS);
```
NIT unnecessary new line

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/utils/ToleranceUtilsTest.java:
```diff
@@ -24,75 +19,44 @@
+    double distanceToIntersection = route.distance();
+    LineString lineString = LineString.fromPolyline(route.geometry(), Constants.PRECISION_6);
     Point closePoint
       = TurfMeasurement.along(lineString, distanceToIntersection, TurfConstants.UNIT_METERS);
```
NIT unnecessary new line

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteLegProgressTest.java:
```diff
@@ -17,77 +12,46 @@
-      .build();
+    routeProgress = routeProgress.toBuilder().stepIndex(lastStepIndex).build();
+
     assertNull(routeProgress.currentLegProgress().upComingStep());
```
Should be in the `Arrange` part 👀 `upComingStep_returnsNull` is telling us what's the `Act` here `upComingStep`.
Also 👀 https://github.com/mapbox/mapbox-navigation-android/pull/812#discussion_r182119363

This applies to other modified tests included in this PR (`RouteProgressTest` and `RouteStepProgressTest`).

Guardiola31337 commented on libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/TestRouteBuilder.java:
```diff
@@ -0,0 +1,17 @@
+
+import static okhttp3.internal.Util.UTF_8;
+
+class TestRouteBuilder {
```
👀 https://github.com/mapbox/mapbox-navigation-android/pull/812#discussion_r182375873

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/TestRouteProgressBuilder.java:
```diff
@@ -0,0 +1,77 @@
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.findCurrentIntersection;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.findUpcomingIntersection;
+
+class TestRouteProgressBuilder {
```
👀 https://github.com/mapbox/mapbox-navigation-android/pull/812#discussion_r182375873

Guardiola31337 commented on libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/TestRouteProgressBuilder.java:
```diff
@@ -0,0 +1,72 @@
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.findCurrentIntersection;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.findUpcomingIntersection;
+
+class TestRouteProgressBuilder {
```
👀 https://github.com/mapbox/mapbox-navigation-android/pull/812#discussion_r182375873

Guardiola31337 commented on libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/TestRouteBuilder.java:
```diff
@@ -0,0 +1,33 @@
+
+import static okhttp3.internal.Util.UTF_8;
+
+class TestRouteBuilder {
```
👀 https://github.com/mapbox/mapbox-navigation-android/pull/812#discussion_r182375873
